### PR TITLE
Consider artifact repositories backed by S3 secure

### LINKED ISSUE #42732: [build] assertRepositoryURIUsesHttps breaks s3 repositories
When building a custom plugin using the build-tools jar, we get the following exception when building the 6.8.0 version:

```
Caused by: java.net.MalformedURLException: unknown protocol: s3
        at java_net_URI$toURL.call(Unknown Source)
        at org.elasticsearch.gradle.BuildPlugin.assertRepositoryURIUsesHttps(BuildPlugin.groovy:598)
        at jdk.internal.reflect.GeneratedMethodAccessor342.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.elasticsearch.gradle.BuildPlugin$_configureRepositories_closure11.doCall(BuildPlugin.groovy:567)
        at jdk.internal.reflect.GeneratedMethodAccessor341.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.util.ClosureBackedAction.execute(ClosureBackedAction.java:70)
        at org.gradle.util.ConfigureUtil.configureTarget(ConfigureUtil.java:154)
        at org.gradle.util.ConfigureUtil.configure(ConfigureUtil.java:105)
        at org.gradle.util.ConfigureUtil$WrappedConfigureAction.execute(ConfigureUtil.java:166)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1$1.run(DefaultCollectionCallbackActionDecorator.java:100)
        at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:58)
        at org.gradle.api.internal.DefaultCollectionCallbackActionDecorator$BuildOperationEmittingAction$1.run(DefaultCollectionCallbackActionDecorator.java:97)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
        ... 173 more
```

This is because our build.gradle file has a repositories block like so:

```
    repositories {
        maven {
            url "s3://...
       }
    }
```

### PR DESCRIPTION
**Pull request by jasontedor**
Since credentials are required to access such a repository, and these repositories are accessed over an encrypted protocol (https), this commit adds support to consider S3-backed artifact repositories as secure. Additionally, we add tests for this functionality.

Closes #42732

### CODE REVIEW

**Comment by synhershko** [comment]
For completeness, you might want to consider adding gcs as well, see https://docs.gradle.org/current/userguide/repository_types.html#sub:supported_transport_protocols

**Comment by jasontedor** [comment]
@elasticmachine run elasticsearch-ci/2

File: buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java [hunk 1]
```diff
@@ -36,4 +39,20 @@ public void testFailingDockerVersions() {
         BuildPlugin.checkDockerVersionRecent("Docker version 17.04.0, build e68fc7a");
     }
 
+    @Test(expected = GradleException.class)
+    public void testRepositoryURIThatUsesHttpScheme() throws URISyntaxException {
+        final URI uri = new URI("http://s3.amazonaws.com/artifacts.elastic.co/maven");
+        BuildPlugin.assertRepositoryURIIsSecure("test", "test", uri);
+    }
+
+    public void testRepositoryURIThatUsesHttpsScheme() throws URISyntaxException {
```
Comment (pugnascotia) [inline buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java]: Can we add a test for `file:` as well?

**Review by pugnascotia — APPROVED** [review]

File: buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java (hunk 1 above)
Comment (jasontedor) [reply buildSrc/src/test/groovy/org/elasticsearch/gradle/BuildPluginTests.java]: I pushed f028816b2cf3e7f37e9e4be77e05fd9d94372116.

**Comment by jasontedor** [comment]
Thanks for reviewing @pugnascotia.

**Comment by jasontedor** [comment]
@synhershko We'll add GCS-backed repositories if the need arises, but I'd rather not add for the sake of adding (despite how easy it looks, it still adds to maintenance and testing requirements).

**Comment by synhershko** [comment]
@jasontedor sure that makes sense. Thanks for the super quick fix!

**Pull request merged**